extensions/authentication-passport/README.md

```diff
@@ -0,0 +1,245 @@
+
+_Important: We strongly suggest that users understand LoopBack's
+[authentication system](https://loopback.io/doc/en/lb4/Loopback-component-authentication.html)
+before using this module_
+
+This is an adapter module created for plugging in
+[`passport`](https://www.npmjs.com/package/passport) based strategies to the
+authentication system in `@loopback/authentication@2.x`.
+
+## Installation
+
+```sh
+npm i @loopback/authentication-passport --save
+```
+
+## Background
+
+`@loopback/authentication@2.x` allows users to register authentication
+strategies that implement the interface
+[`AuthenticationStrategy`](https://apidocs.strongloop.com/@loopback%2fdocs/authentication.html#AuthenticationStrategy)
+
+Since `AuthenticationStrategy` describes a strategy with different contracts
+than the passport
+[`Strategy`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/passport/index.d.ts#L79),
+and we'd like to support the existing 500+ community passport strategies, an
+**adapter class** is created in this package to convert a passport strategy to
+the one that LoopBack 4 authentication system wants.
+
+## Usage
+
+### Simple Usage
+
+1. Create an instance of the passport strategy
+
+Taking the basic strategy exported from
+[`passport-http`](https://github.com/jaredhanson/passport-http) as an example,
+first create an instance of the basic strategy with your `verify` function.
+
+```ts
+// Create a file named `my-basic-auth-strategy.ts` to define your strategy below
+
+import {BasicStrategy} from 'passport-http';
+
+function verify(username: string, password: string, cb: Function) {
+  users.find(username, password, cb);
+}
+const basicStrategy = new BasicStrategy(verify);
+```
+
+It's a similar configuration as you do when adding a strategy to a `passport` by
+calling `passport.use()`.
+
+2. Apply the adapter to the strategy
+
+```ts
+// In file 'my-basic-auth-strategy.ts'
+import {BasicStrategy} from 'passport-http';
+
+function verify(username: string, password: string, cb: Function) {
+  users.find(username, password, cb);
+}
+const basicStrategy = new BasicStrategy(verify);
+
+// Apply the adapter
+export const AUTH_STRATEGY_NAME = 'basic';
+export const basicAuthStrategy = new StrategyAdapter(
+  // The configured basic strategy instance
+  basicStrategy,
+  // Give the strategy a name
+  // You'd better define your strategy name as a constant, like
+  // `const AUTH_STRATEGY_NAME = 'basic'`.
+  // You will need to decorate the APIs later with the same name.
+  AUTH_STRATEGY_NAME,
+);
+```
+
+3. Register(bind) the strategy to app
+
+```ts
+import {Application, CoreTags} from '@loopback/core';
+import {AuthenticationBindings} from '@loopback/authentication';
+import {basicAuthStrategy} from './my-basic-auth-strategy';
+
+app
+  .bind('authentication.strategies.basicAuthStrategy')
+  .to(basicAuthStrategy)
+  .tag({
+    [CoreTags.EXTENSION_FOR]:
+      AuthenticationBindings.AUTHENTICATION_STRATEGY_EXTENSION_POINT_NAME,
+  });
```

Comment:
Do we have a helper binding template function for this, such as `asAuthenticationStrategy`?

Reply:
@raymondfeng yep we have `registerAuthenticationStrategy()`, while it doesn't work for this use case, because the `basicAuthStrategy` here is a class instance, not a class. (`registerAuthenticationStrategy()` takes in a class)

Reply:
I'm referring to https://loopback.io/doc/en/lb4/Binding.html#apply-a-template-function